File: wrappers/julia/ppr_path.jl
# A julia wrapper for ppr_path
# A - sparse matrix representation of graph
# alpha - value of alpha
# eps - value of epsilon
# rho - value of rho
# seedids,nseedids - the set of indices for seeds
# xlength - the max number of ids in the solution vector
# xids, actual_length - the solution vector
# values - the pagerank value vector for xids (already sorted in decreasing order)
function ppr_path{T}(A::SparseMatrixCSC{T,Int64},alpha::Float64,
                          eps::Float64,rho::Float64,seedids,nseedids,xlength)
    offset = 1;
    n=A.n
    xids=zeros(Int64,xlength);
    actual_length=ccall((:ppr_path64,"../../lib/graph_lib_test/libgraph"),Int64,
                        (Int64,Ptr{Int64},Ptr{Int64},Int64,Cdouble,Cdouble,Cdouble,
                        Ptr{Int64},Int64,Ptr{Int64},Int64),n,A.colptr,A.rowval,
                        offset,alpha,eps,rho,seedids,nseedids,xids,xlength); 
    xids=xids[1:actual_length];
    return actual_length,xids
end

File: wrappers/matlab/ppr_path.m

% A matlab wrapper for aclpagerank
% A - sparse matrix representation of graph
% alpha - value of alpha
% eps - value of epsilon
% seedids,nseedids - the set of indices for seeds
% maxsteps - the max number of steps
% xlength - the max number of ids in the solution vector
% xids, actual_length - the solution vector
% values - the pagerank vector for xids (already sort decreasing order)
function [actual_length,xids]=ppr_path(A,seedids, ...
                                        alpha,eps,rho,xlength)
[ajPtr,aiPtr,~]=sparse_arrays_ptr(A);
[n,~]=size(A);
[nseedids,~]=size(seedids);
x=zeros(xlength,1);
switch computer
    case {'PCWIN64','GLNXA64','MACI64'}
        indtype = 'int64Ptr';
    case {'PCWIN','GLNX86'}
        indtype = 'uint32Ptr';
    otherwise
        error('Unsupported system');
end
xPtr = libpointer(indtype,x);
seedPtr = libpointer(indtype,seedids);
loadlibrary('../../lib/graph_lib_test/libgraph','../../lib/include/ppr_path_c_interface.h')
if strcmp(indtype,'int64Ptr')
    actual_length = calllib('libgraph','ppr_path64',n,aiPtr, ...
                            ajPtr,0,alpha,eps,rho,seedPtr,nseedids, ...
                            xPtr,xlength);
else
    actual_length = calllib('libgraph','ppr_path32',n,aiPtr, ...
                            ajPtr,0,alpha,eps,rho,seedPtr,nseedids, ...
                            xPtr,xlength);
end    
xids=get(xPtr,'Value');
xids=xids(1:actual_length);
unloadlibrary libgraph;